lib/modules: optimize byName

pennae opened the pull request:
###### Motivation for this change

the foldl is equivalent to a zip with concat. list concatenation in nix
is an O(n) operation, which makes this operation extremely inefficient
when large numbers of modules are involved.

this change reduces the number of list elements by 7 million on the
system used to write this, total memory spent on lists by 58MB, and
total memory allocated on the GC heap by almost 100MB (with a similar
reduction in GC heap size). it's also slightly faster.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

pennae committed:
lib/modules: optimize byName

the foldl is equivalent to a zip with concat. list concatenation in nix
is an O(n) operation, which makes this operation extremely inefficient
when large numbers of modules are involved.

this change reduces the number of list elements by 7 million on the
system used to write this, total memory spent on lists by 58MB, and
total memory allocated on the GC heap by almost 100MB (with a similar
reduction in GC heap size). it's also slightly faster.

roberth commented on lib/modules.nix:
```diff
@@ -442,7 +443,8 @@ rec {
+          (map (module:
               if !(builtins.isAttrs module.${attr}) then
                 throw ''
                   You're trying to declare a value of type `${builtins.typeOf module.${attr}}'
```
This function is used on both `options` and `config`, yet the message is the same. This can't be right.
Perhaps the function should be duplicated, specializing for `config` and `options` respectively.

pennae replied:
turning zipAttrsWith into a primop is massively more efficient and benefits other users too (eg recursiveUpdate). we've made a prototype implementation of `__zipAttrsWith` and have seen another sizable decrease in resource usage on system builds :)

roberth reviewed (commented):
In the evaluation report (see commit status), envs, calls and values are up.

You could probably reduce envs and calls by specializing as suggested, then inlining `zipAttrsWith` to see if it's doing things you don't need it to.

I don't understand why values are up. What do you think?

If you don't want to delve deeper, I'll accept that, but it's probably worth doing, if you haven't already done so.

Is `nix-instantiate -A nixosTests.acme` unchanged? (as an example)

pennae commented:
values are up because what previously never left foldl' (and thus was never accounted in the eval reports) now makes its way into nix (via the result of map), back into builtins and out again a couple time (for attrNames in zipAttrs, map, catAttrs, concatLists, listToAttrs). the values were always there, they just weren't counted.

specializing and inlining zipAttrsWith doesn't seem worth the code duplication—what might be worth doing though is removing zipAttrsWithNames (only used to define zipAttrsWith) and turning zipAttrsWith into a builtin. we'd still have `map` to contend with as far as value count goes, but it's well worth not doing n² builds on multiple lists of 1000 elements or more.

we've tested this on our full system config, the only notable change in nix-diff was nixos version info. haven't run any tests though.

pennae committed:
lib/attrsets: use builtins.zipAttrsWith if available

pennae commented:
proposed a nix primop for zipAttrsWith here: https://github.com/NixOS/nix/pull/5830

pennae committed:
lib/modules: extract multiply-used value in byName

module.${attr} is used at least twice, so it must be evaluated at least
twice (and since it's a function argument, be turned into a thunk
twice).

pennae commented:
zipAttrsWith has been merged in nix :tada:

pennae commented:
we've run tests for acme, unbound, systemd and containers to make sure system eval results are the same. they were for acme, unbound and systemd, the container tests differed only in a reference to the nixpkgs source location.

@roberth can we consider this verified?

roberth commented:
Relevant stats:
final heap size reduction
ofborg: -1.63%
nix master: -5.25%

total allocations reduction
ofborg: -0.61%
nix master: -3.25%

time: no reliable data

pennae commented:
not currently planning on adding any more, but adding crosschecks might be a good idea if we add more. maybe have a way to eval (almost) all of nixpkgs with and without the polyfills, then compare the results. :thinking:
